File: DataInjectorIOS/DataInjector/Classes/utility/InjectorDataDetector.swift
//
//  InjectorDataDetector.swift
//  DataInjector Pod
//
//  Library utility: detect data types
//  Determine a data type from an object or string
//

import Foundation

/// A utility to help parse data types from a string used in conditions or mappers
public class InjectorDataDetector {
    
    // ---
    // MARK: Initialization
    // ---
    
    private init() {
    }
    
    
    // ---
    // MARK: Detection
    // ---
    
    public static func detectFromObject(_ object: Any?) -> InjectorDataType {
        if object is String {
            return detectFromString(object as! String)
        } else if object is Float || object is Double {
            return .decimalNumber
        } else if object is Int {
            return .number
        } else if object is Bool {
            return .boolean
        } else if object == nil || object is NSNull {
            return .empty
        }
        return .unknown
    }
    
    public static func detectFromString(_ value: String, start: String.Index? = nil) -> InjectorDataType {
        var offsetString = value
        if start != nil {
            offsetString = value.substring(from: start!)
        }
        if offsetString.hasPrefix("'") || offsetString.hasPrefix("\"") {
            return .string
        } else if offsetString.hasPrefix("@.") {
            return .subReference
        } else if offsetString.hasPrefix("@") {
            return .reference
        } else if offsetString.hasPrefix("true") || offsetString.hasPrefix("false") {
            return .boolean
        } else if offsetString.hasPrefix("empty") {
            return .empty
        } else if offsetString.count > 0 && (offsetString[offsetString.startIndex] >= "0" && offsetString[offsetString.startIndex] <= "9") {
            return containsDot(numberString: offsetString) ? .decimalNumber : .number
        } else if offsetString.count > 1 && offsetString[offsetString.startIndex] == "-" && (offsetString[offsetString.index(after: offsetString.startIndex)] >= "0" && offsetString[offsetString.index(after: offsetString.startIndex)] <= "9") {
            return containsDot(numberString: offsetString) ? .decimalNumber : .number
        }
        return .string
    }
    
    public static func endOfTypeString(type: InjectorDataType, value: String, start: String.Index? = nil) -> String.Index? {
        var checkStart = start ?? value.startIndex
        var offsetString = value.substring(from: checkStart)
        if type == .string && (offsetString.hasPrefix("'") || offsetString.hasPrefix("\"")) {
            let findEndChar = offsetString[offsetString.startIndex]
            checkStart = value.index(after: checkStart)
            for index in value.indices[checkStart..<value.endIndex] {
                if value[index] == findEndChar {
                    return value.index(after: index)
                }
            }
        } else if type == .number || type == .decimalNumber {
            checkStart = value.index(after: checkStart)
            for index in value.indices[checkStart..<value.endIndex] {
                if !(value[index] >= "0" && value[index] <= "9") && value[index] != "." {
                    return index
                }
            }
        } else if type == .boolean {
            if offsetString.hasPrefix("true") {
                return value.index(checkStart, offsetBy: 4)
            } else if offsetString.hasPrefix("false") {
                return value.index(checkStart, offsetBy: 5)
            }
        } else if type == .empty {
            return value.index(checkStart, offsetBy: 5)
        }
        return nil
    }

    
    // ---
    // MARK: Helper
    // ---
    
    private static func containsDot(numberString: String) -> Bool {
        for index in numberString.indices {
            let chr = numberString[index]
            if chr == "." {
                return true
            }
            if !(chr >= "0" && chr <= "9") && chr != "-" {
                return false
            }
        }
        return false
    }

}

File: DataInjectorAndroid/DataInjectorLib/src/main/java/com/crescentflare/datainjector/utility/InjectorDataDetector.java
package com.crescentflare.datainjector.utility;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Data injector utility: detect data types
 * Determine a data type from an object or string
 */
public class InjectorDataDetector
{
    // ---
    // Initialization
    // ---

    private InjectorDataDetector()
    {
    }


    // ---
    // Detection
    // ---

    public static InjectorDataType detectFromObject(Object object)
    {
        if (object instanceof String)
        {
            return detectFromString((String)object);
        }
        else if (object instanceof Double || object instanceof Float)
        {
            return InjectorDataType.DecimalNumber;
        }
        else if (object instanceof Integer)
        {
            return InjectorDataType.Number;
        }
        else if (object instanceof Boolean)
        {
            return InjectorDataType.Boolean;
        }
        else if (object == null)
        {
            return InjectorDataType.Empty;
        }
        return InjectorDataType.Unknown;
    }

    public static InjectorDataType detectFromString(String value)
    {
        return detectFromString(value, 0);
    }

    public static InjectorDataType detectFromString(String value, int start)
    {
        if (start < 0)
        {
            return InjectorDataType.Unknown;
        }
        if (value.startsWith("'", start) || value.startsWith("\"", start))
        {
            return InjectorDataType.String;
        }
        else if (value.startsWith("@.", start))
        {
            return InjectorDataType.SubReference;
        }
        else if (value.startsWith("@", start))
        {
            return InjectorDataType.Reference;
        }
        else if (value.startsWith("true", start) || value.startsWith("false", start))
        {
            return InjectorDataType.Boolean;
        }
        else if (value.startsWith("empty", start))
        {
            return InjectorDataType.Empty;
        }
        else if (value.length() > start && value.charAt(start) >= '0' && value.charAt(start) <= '9')
        {
            return containsDot(value) ? InjectorDataType.DecimalNumber : InjectorDataType.Number;
        }
        else if (value.length() > start + 1 && value.charAt(start) == '-' && value.charAt(start + 1) >= '0' && value.charAt(start + 1) <= '9')
        {
            return containsDot(value) ? InjectorDataType.DecimalNumber : InjectorDataType.Number;
        }
        return InjectorDataType.String;
    }

    public static int endOfTypeTypeString(InjectorDataType type, String value)
    {
        return endOfTypeTypeString(type, value, 0);
    }

    public static int endOfTypeTypeString(InjectorDataType type, String value, int start)
    {
        if (start < 0)
        {
            return -1;
        }
        if (type == InjectorDataType.String && (value.startsWith("'", start) || value.startsWith("\"", start)))
        {
            char findEndChar = value.charAt(start);
            int len = value.length();
            for (int i = start + 1; i < len; i++)
            {
                if (value.charAt(i) == findEndChar)
                {
                    return i + 1;
                }
            }
        }
        else if (type == InjectorDataType.Number || type == InjectorDataType.DecimalNumber)
        {
            int len = value.length();
            for (int i = start + 1; i < len; i++)
            {
                char chr = value.charAt(i);
                if (!(chr >= '0' && chr <= '9') && chr != '.')
                {
                    return i;
                }
            }
        }
        else if (type == InjectorDataType.Boolean)
        {
            if (value.startsWith("true", start))
            {
                return start + 4;
            }
            else if (value.startsWith("false", start))
            {
                return start + 5;
            }
        }
        else if (type == InjectorDataType.Empty)
        {
            return start + 5;
        }
        return -1;
    }


    // ---
    // Helper
    // ---

    private static boolean containsDot(String numberString)
    {
        int len = numberString.length();
        for (int i = 0; i < len; i++)
        {
            char chr = numberString.charAt(i);
            if (chr == '.')
            {
                return true;
            }
            if (!(chr >= '0' && chr <= '9') && chr != '-')
            {
                return false;
            }
        }
        return false;
    }
}
